### ISSUE TITLE
May i know when new composer package version will release? I am getting create_function() deprecated notice in the PHP 7.2

### ISSUE THREAD

**Issue by pmbaldha** [phpseclib/phpseclib#1293]
I am getting create_function() deprecated notice in the PHP 7.2. You have already replaced create_function() with anonymous function in the master version.

**Comment by terrafrost** [comment]
Can you give me some sample code or maybe a file name / line number?

I ask because this should have been fixed 11 months ago:

https://github.com/phpseclib/phpseclib/commit/98d46db7c51cf8e4f8648b383d28db78dd22d055

phpseclib 1.0.8 / 2.0.7 introduced this change:

https://github.com/phpseclib/phpseclib/releases/tag/1.0.8
https://github.com/phpseclib/phpseclib/releases/tag/2.0.7

phpseclib is passing all the unit tests on travis-ci on 7.2:

https://travis-ci.org/phpseclib/phpseclib/builds/418510260
https://travis-ci.org/phpseclib/phpseclib/builds/418510293

A deprecated notice would cause the unit tests to fail...

**Comment by pmbaldha** [comment]
@terrafrost Thank you very much for your reply.

I have just checked phpseclib 2.0.11. There are still the create_function exist at line no.  2610 in the phpseclib/Crypt/Base.php file. This line as below:

`return create_function('$_action, &$self, $_text', $init_crypt . 'if ($_action == "encrypt") { ' . $encrypt . ' } else { ' . $decrypt . ' }');`

Thanks you

**Comment by markkap** [comment]
the usage of `eval` as a replacement of `create_function` is not very trust inspiring especially in a security oriented library. Just saying.... 

(randomly got to this issue while trying to assess the capabilities of the library)

**Comment by terrafrost** [comment]
`eval` is only a security issue if user input is utilized in the string that's passed to `eval`. If you - or anyone - can find an exploit - specifically in phpseclib's use of `eval` - I will pay said individual $500 USD via PayPal. (this offer does not extend to timing attacks)

I did `grep -r "eval(" .`  in the vendor directory of a work project of mine and got lots of results back. Symfony uses it in a number of places, including here:

https://github.com/symfony/http-kernel/blob/master/HttpCache/HttpCache.php#L576

I'm not going to list every project that uses `eval()` but sufficient to say projects bigger than phpseclib use it.

If phpseclib were to stop using eval than the 35% speed increase that #1143 and #65 introduced go away. There is no other practical way to get that kind of speed up with pure-PHP code for those kinds of things. If you can think of a way comparable speeds can be obtained using pure-PHP code and without the use of `eval` I'll pay you $500 USD PayPal.

Note that the offer to pay $500 USD PayPal is not a permanent offer. It depends on my job situation / financial situation at the time. And I'm not going to pay out $500 USD twice because my own personal financial resources are somewhat constrained.

In summary, `eval` isn't inherently bad. It can be used in insecure ways but it is not inherently insecure and sometimes it just offers functionality that you can't get through any other mechanism (feel free to prove me wrong). If you're gonna write phpseclib off as a crap product because phpseclib uses `eval` you might as well write also write it off because it was written in PHP. I mean, in the greater ecosystem, PHP has a pretty abysmal reputation, too, that's largely not based in fact.

**Comment by markkap** [comment]
It is not about having an actual exploit, it is about security in depth. In the code displayed here if I can make $init_crypt to be something like `}; exec("rm -f /") //;` the server is mine. It might be that current code paths prevent such a thing from happening, but some future change might introduce such an execution path.

Even despised software (from a security POV) like wordpress successfully avoids using eval in PHP.

The modern way to do what you are trying to do is probably by using a closure aka anonymous function, which is the true replacement to create_function.

**Comment by terrafrost** [comment]
> It might be that current code paths prevent such a thing from happening, but some future change might introduce such an execution path.

In most cases I would agree but in this case we'd be taking a significant performance hit.

> The modern way to do what you are trying to do is probably by using a closure aka anonymous function, which is the true replacement to create_function.

`eval` is used in this file:

https://github.com/phpseclib/phpseclib/blob/master/phpseclib/Math/BigInteger/Engines/PHP/Reductions/EvalBarrett.php

`eval` is being used to create an anonymous function. So you're proposal is to eliminate this `eval` created anonymous function with a non-`eval` created anonymous function?

In this particular case the 35% speedup is obtained through the dynamic generation of code. Instead of having a generic function to give you x mod y, for any y, this is a custom function that gives x mod y for a specific value of y.

Consider this code (using the master branch):

```php
<?php
require __DIR__ . '/vendor/autoload.php';

use phpseclib\Math\BigInteger;
use phpseclib\Crypt\RSA;

BigInteger::setEngine('PHP64');

$public = new RSA();
$public->load('-----BEGIN PUBLIC KEY-----
MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCqGKukO1De7zhZj6+H0qtjTkVxwTCpvKe4eCZ0
FPqri0cb2JZfXJ/DgYSF6vUpwmJG8wVQZKjeGcjDOL5UlsuusFncCzWBQ7RKNUSesmQRMSGkVb1/
3j+skZ6UtW+5u09lHNsj6tQ51s1SPrCBkedbNf0Tp0GbMJDyR4e9T04ZZwIDAQAB
-----END PUBLIC KEY-----');
$public->encrypt('zzz');
```

EvalBarrett.php creates a 40kb bespoke function based o the modulo. I won't post the function in-line because it's long but here it is on pastebin.com:

https://pastebin.com/Hi58iL1A

Now consider this code (from multiplyHelper in phpseclib/Math/BigInteger/Engines/PHP.php):

```php
        $carry = 0;
        for ($j = 0; $j < $x_length; ++$j) { // ie. $i = 0
            $temp = $x_value[$j] * $y_value[0] + $carry; // $product_value[$k] == 0
            $carry = static::BASE === 26 ? intval($temp / 0x4000000) : ($temp >> 31);
            $product_value[$j] = (int) ($temp - static::BASE_FULL * $carry);
        }
```
In the anonymous function that was created through `eval` `$x` is hard-coded. The for loop is eliminated. The `++$j` is not done. `$x_value[$j]` is replaced with the actual int for `$j`. So for `$x_value[0]` we do `1961628378` instead. For `$x_value[1]` we do `1225269599`. And instead of doing the `static::BASE === 26` it just does the correct division operation automatically because the bespoke function that was created knew - at runtime - whether or not PHP was 32-bit or 64-bit.

And instead of doing `static::BASE_FULL` we do `2147483648`.

So there's loop unrolling, array dereferencing, and the elimination of if conditions.

It's basically a whole bunch of micro-optimized code that is COMPLETELY dependent upon the modulo. You cannot write a general purpose modulo method with loop unrolling but if you assume the modulo to be fixed then you can.

**Comment by terrafrost** [comment]
TLDR I submit that there is ZERO way to achieve the 35% speedups the a-fore mentioned pull requests have done without the use of `eval`. And like I said, if you *can* think of a way to do so WITHOUT `eval`, I will pay you $500 USD via PayPal. I will be absolutely flabbergasted.

Sure, the fact that `eval` is being used makes it easier for vulnerabilities to be introduced, but really don't expect I'll ever get a PR for the lines that make use of `eval` because that code is a nightmare to wrap your head around as is. It's basically PHP code that generates PHP code and there's not really standards for that.

If you trust my judgement so little that you think I might still, none-the-less, hypothetically speaking, accept a PR to those files without _extreme_ vetting then I have to wonder why you would assume that the code doesn't already have intentional backdoors placed within it. Ken Thompson's seminal talk _[Reflections on Trusting Trust](https://www.archive.ece.cmu.edu/~ganger/712.fall02/papers/p761-thompson.pdf)_ addresses this some. Sometimes you just have to assume that the people who are creating this shit know what they're doing and just as you have to take it on faith that `gcc` doesn't have a backdoor in it (per the talk) at some level you're gonna have to take it on faith that I know what I'm doing. And if you can't take it on faith then don't use this library. Write your own SSH library or RSA library or whatever it is that you're using phpseclib for.

**Comment by terrafrost** [comment]
Here's an alternate explanation that [/u/kemmeta](https://www.reddit.com/user/kemmeta) on [/r/PHP](https://www.reddit.com/r/PHP/) posted:

> The functions that are being dynamically created aren't just mode specific - they're key specific.
> 
> The function that the `eval` generates for AES with a key / IV of "zzz" / "zzz" is different than the function that it generates for "aaa" / "aaa" and the way it's generating these key specific functions it's able to achieve a significant speedup.
> 
> A good way to think about it is...  what about be faster? This?:
> 
>     $a + $b + $c + $d + ($e ? 50 : 80)
> 
> ...or (if you assume every variable but `$a` is known in advance) this?:
> 
>     $a + 167
> 
> phpseclib is using `eval` to make these kinds of parameter specific optimizations.
> 
> Loop unrolling is also employed:
> 
>     for ($i = 0; $i < $length; $i++) {
>         for ($j = 0; $j < 100; $j++) {
>             // do stuff
>        }
>     }
> 
> You can eliminate the inner for loop with 100 "do stuff" calls.
> 
> It's all a bunch of micro-optimizations but when you're potentially dealing with hundreds of megabytes of data it can make a big difference. In phpseclib's case these micro-optimizations result in a 35% speedup I think the post was saying.

Source: https://www.reddit.com/r/PHP/comments/9mfetd/the_land_where_php_uses_eval/e7ef4x7/

Upon review, [/u/SaraMG](https://www.reddit.com/user/SaraMG), a core PHP developer, said the following:

> I'm going to hold onto this library as a benchmark for engine work and as the first arguably legitimate use of eval I've ever seen in almost 20 years of using this language.

Source: https://www.reddit.com/r/PHP/comments/9mfetd/the_land_where_php_uses_eval/e7fesdw/

The /r/PHP discussion resulted in a series of tweets on the subject:

https://twitter.com/SaraMG/status/1049477643077337088

One of the people who responded to the tweet [said of my responses](https://twitter.com/ramsey/status/1049481077549555713), here, "[TerraFrost] _is not being very nice to the OP about it_".

All I'll say on that subject is that I have invested a lot of time into this library and if I feel it's being subject to unfair criticism that I can take that personally. If I didn't take it personally, I wouldn't have offered the bounty that I did. [/u/kemmeta](https://www.reddit.com/user/kemmeta), however, was able to be more impartial than I was in their explanation so I am grateful to them.

The main reason I'm documenting all this is because of #1300. I figure this bug report / issue can be the canonical source of info on phpseclib's use of `eval`.

**Comment by bantu** [comment]
> I am getting create_function() deprecated notice in the PHP 7.2.

@pmbaldha Where are you getting that? Static code analysis?

**Comment by bantu** [comment]
@terrafrost phpseclib will function without EvalBarrett, so an option would be moving it to its own composer package, which can be depended on optionally. The "best practice" would be to install the GMP module anyway.

**Comment by petrich** [comment]
Is the $500 USD PayPal offer still valid? :zany_face:

**Comment by terrafrost** [comment]
@petrich - I guess it'd depend on the solution you came up with lol. You are the father of the current `eval` approach so if anyone could do it it'd be you!

I guess I'd say: let's see what you've come up with and then we can talk!

That said, the offer was kinda made in the heat of the moment - at the time I felt like people were unjustly ganging up on me and it was kinda an attempt to get people to back off lol. Like maybe with the cash offer people would be motivated to try to achieve similar performance boosts without the use of `eval` and upon failing they'd have a change of heart and thus I'd feel less ganged up on lol. And ultimately SaraMG, a core PHP dev I have a ton of respect for, did seem to have a change of heart. Would she have been motivated to look into it enough to change her mind had it not been for my $500 cash offer at the time? Who knows lol.

But regardless, I think you did an awesome job when you came up with that technique and if you have a new approach it could very well be worth it's weight in gold and I'd love to see it!

**Comment by petrich** [comment]
Hi Terra :-) I'm happy phpseclib is still very alive and well. :+1: 

The idea is:

(pseudo code)
```php
$include_file = "( vfs:// | /dev/shm | /tmp )/{$inline_code_hash}.inc.php";  
// ( vfs:// --> if possible, a string based virtual filesystem in pure php, based on the php stream_* functions, so we not have to deal with real files )

file_put_content($include_file, "<?php function fnc_{$inline_code_hash}() { $inline_code;} ?>");
require_once "$include_file";
```

Technicaly... this would replace `eval("function fnc_{$inline_code_hash}() { $inline_code;}")` 

:smiley: 

I remember, this even was my first idea... before i quickly realized that `create_function()` or `eval()` would be better/faster/easier... executing on-the-fly generated phph code.

The security issue, of course, would still remain. Because it doese'nt wether the generated php code is executed via `eval()` or via `require_once` 

Only just in case that the PHP Group will disable `eval()`  anytime  (but i think not) ... 
above `require_once` idea could be an alternative implemention of `inline_crypt` without using `eval()`.

btw: Imho... phpseclibs use of `eval()` for inline_crpyt is ultrasafe. I was never worry about this.

**Comment by terrafrost** [comment]
It's an interesting idea! Same basic idea as `eval` but just kinda executed differently.

I feel like it could be more fragile than `eval()` however? Like wouldn't `allow_url_include` settings potentially break it? I wrote this script to test the general idea out and it didn't work:

```php
<?php

$fp = fopen('php://memory', 'w');
fputs($fp, '<?php echo "hello, world!";');
rewind($fp);
fclose($fp);
include('php://memory');
```
Like I'd expect that to output "hello, world!" but it didn't. I tried it with and without the `fclose()`. Am I doing something wrong?

> btw: Imho... phpseclibs use of `eval()` for inline_crypt is ultrasafe. I was never worry about this.

Yah I agree. It's not putting user input into `eval()`. Even if you do `$cipher->setKey($_GET['whatever'])` it's not passing that straight into `eval()` - it's unpacking that, among other things.

**Comment by petrich** [comment]
Hi Terra :handshake: 

unfortunately, `php://memory` will not work for `include` for some reasons. 

Main reason is (beside `allow_url_include`): Access to the content is only possible via the stream resource `$fp` ... not via the path `php://memory` ... and `include` works only with paths , not with stream resources `$fp`.

What will work is:
```php 
include('data://text/plain;base64,'.base64_encode('<?php echo "hello, world!";'));
```

Problem here: The internal stream wrapper `data://` runs with `include` only if php.ini: `allow_url_include=1` ... which is rare ... and also flaged as deprecated.
So unfortunately... `data://` is also no choice.

But, luckily :-)  custom memory stream wrappers, written in pure php, will work with `include` independently from the `allow_url_include` setting.

Just for proof of concept... i  picked up a random php projekt which already implements a memory based virtual file system written in pure php.

So this code works... fast... and all in-memory. I think the code is self explaining:

```php
require __DIR__ . '/vendor/autoload.php';

// https://github.com/adlawson/php-vfs
use Vfs\FileSystem;
use Vfs\Node\Directory;
use Vfs\Node\File;

$fs = FileSystem::factory('vfs://');
$fs->mount();

$vfile = new Directory(['code.php' => new File('<?php echo "Hello, World!\n";')]);
$fs->get('/')->add('inline-crypt', $vfile);

include 'vfs://inline-crypt/code.php'; 

```

Only just for the record ... as an alternative to eval()... if needed someday.

**Comment by terrafrost** [comment]
It's definitely an interesting idea!

I guess my thoughts are: it's prob not gonna be significantly faster than `eval` and would require a large rewrite of a lot of code, which could introduce new problems.

If `eval` does get deprecated it's a great alternative tho!
